annotations/views/rest_views.py

```diff
@@ -476,25 +480,87 @@ def create(self, request, *args, **kwargs):
+            results = map(_relabel, [c.data for c in concepts])
+
+            def _parseViaf(self, uri, new_concepts):
+                data = urllib2.urlopen(uri)
+                tree = e.parse(data)
+                root = tree.getroot()
+                namespace = {'foaf': 'http://xmlns.com/foaf/0.1/', 'ns1': 'http://viaf.org/viaf/terms#', 'owl': 'http://www.w3.org/2002/07/owl#', 'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#', 'void': 'http://rdfs.org/ns/void#'}
+                find = e.XPath("//ns1:text", namespaces = namespace)
+                concept_name = find(root)[0].text
+                concept_description = find(root)[0].text
+
+                dic = {
+                    'label': concept_name,
+                    'desc': concept_description,
+                    'uri': uri,
+                    'auth': "VIAF"}
+                new_concepts.append(dic)
+
+
+            def _parseHps(self, uri, new_concepts):
+                data = urllib2.urlopen(uri)
+                tree = ET.parse(data)
+                root = tree.getroot()
+                namespace = {'hps': 'http://www.digitalhps.org/'} 
+
+                for entry in root.findall('hps:conceptEntry', namespace):
+                    description = entry.find('hps:description', namespace)
+                    name = entry.find('hps:lemma', namespace)
+                    concept_description = description.text
+                    concept_name = name.text
+
+                dic = {
+                    'label': concept_name,
+                    'desc': concept_description,
+                    'uri': uri,
+                    'auth': "Concept Power"}
+                new_concepts.append(dic)
+
+            for result in results:
+                conceptsSet = Set()
+                for ident in result["identities"]: # go through the identities in each result
+                    identSet = Set(ident['concepts'])
+                    diff = identSet.difference(conceptsSet)
+                    conceptsSet.update(diff)
+                result["identities"] = list(conceptsSet) # replace the identities list
+                if result["identities"]:
+                    if result["uri"] in result["identities"]:
+                        result["identities"].remove(result["uri"]) # remove original uri from the list if it exists. 
+                    new_concepts = []
+                    for concept in result["identities"]:
+                        hps = re.search( r'www.digitalhps.org', concept, re.M|re.I)
+                        viaf = re.search( r'viaf.org', concept, re.M|re.I)
+                        if hps:
+                            _parseHps(self, concept, new_concepts)
+                        elif viaf:
+                            _parseViaf(self, concept, new_concepts)
+                    result["identities"] = new_concepts # add the concept data back to the identities list
+            cache.set(q.lower(), results, 60*60*24*14) # two weeks in seconds
```

Comment:
- that should be more like a day
- that should be a configurable value
- I just realized that if we use a cache here, the 'force refresh' checkbox needs to work and considered here (if it is checked, then the cache should not be used). Otherwise, I might search for a concept, not find it, create it in Conceptpower, but it still won't show up.